File: android/src/main/kotlin/com/virtualightning/funky_base/AsyncMethodProcessor.kt
package com.virtualightning.funky_base

import android.content.Context
import android.os.Handler
import io.flutter.plugin.common.MethodChannel
import java.util.*

typealias AsyncCallback = (AsyncResultCallback) -> Unit

/**
 *  Anchor : Create by CimZzz
 *  Time : 2020/03/15 09:04:13
 *  Project : funky_base
 *  Since Version : 1.0.0
 *
 *  异步方法处理元
 *  用来处理异步方法
 */
abstract class AsyncMethodProcessor(methodName: String, context: Context? = null) : BaseMethodProcessor(methodName, context) {
	private val childProcessorList = LinkedList<AsyncProcessor>()
	internal lateinit var handler: Handler

	private fun getAsyncProcessor(arguments: List<Any>?) : AsyncProcessor? {
		var processor: AsyncProcessor? = null
		for(existProcessor in childProcessorList) {
			if(existProcessor.matchArguments(arguments)) {
				processor = existProcessor
				break
			}
		}

		return processor
	}

	internal fun deleteAsyncProcessor(processor: AsyncProcessor) {
		childProcessorList.remove(processor)
	}

	/**
	 * 执行异步处理
	 * 注意在 AsyncCallback 回调执行在主线程，如果需要使用子线程，请在回调中自行处理
	 */
	fun asyncHandle(result: MethodChannel.Result, arguments: List<Any>? = null, callback: AsyncCallback) {
		var processor = getAsyncProcessor(arguments)
		var isNewProcessor = false
		if(processor == null) {
			isNewProcessor = true
			processor = AsyncProcessor(arguments)
			childProcessorList.add(processor)
		}

		processor.addResult(result)

		if(isNewProcessor) {
			val resultCallback = AsyncResultCallback(this, processor)
			callback(resultCallback)
		}
	}
}

class AsyncResultCallback internal constructor(
	private val asyncMethodProcessor: AsyncMethodProcessor,
	private val asyncProcessor: AsyncProcessor
) {
	private var isCompleted = false
	/**
	 * 0 - 成功
	 * 1 - 失败
	 * 2 - 未实现
	 */
	private var completedMode = 0
	private var params: Any? = null
	private var strParams1: String = ""
	private var strParams2: String? = null


	fun doSuccess(data: Any?) {
		if(isCompleted) {
			return
		}
		isCompleted = true
		completedMode = 0
		params = data
		asyncMethodProcessor.handler.obtainMessage(FunkyBasePlugin.ASYNC_COMPLETED, this).sendToTarget()
	}

	fun doError(errorCode: String, errorMessage: String?, errorDetail: Any?) {
		if(isCompleted) {
			return
		}
		isCompleted = true
		completedMode = 1
		params = errorDetail
		strParams1 = errorCode
		strParams2 = errorMessage
		asyncMethodProcessor.handler.obtainMessage(FunkyBasePlugin.ASYNC_COMPLETED, this).sendToTarget()
	}


	fun doNotImplemented() {
		if(isCompleted) {
			return
		}
		isCompleted = true
		completedMode = 2
		asyncMethodProcessor.handler.obtainMessage(FunkyBasePlugin.ASYNC_COMPLETED, this).sendToTarget()
	}

	internal fun perform() {
		when(completedMode) {
			0 -> {
				asyncProcessor.doSuccess(params)
			}
			1 -> {
				asyncProcessor.doError(strParams1, strParams2, params)
			}
			2 -> {
				asyncProcessor.doNotImplemented()
			}
		}
		asyncMethodProcessor.deleteAsyncProcessor(asyncProcessor)
	}
}

/**
 *  Anchor : Create by CimZzz
 *  Time : 2020/03/15 09:08:39
 *  Project : funky_base
 *  Since Version : 1.0.0
 *
 *  异步方法处理子元
 */
internal class AsyncProcessor(private val arguments: List<Any>?) {
	private val resultList = LinkedList<MethodChannel.Result>()

	fun addResult(result : MethodChannel.Result) {
		resultList.add(result)
	}

	fun matchArguments(args: List<Any>?) : Boolean {
		if(arguments == null && args == null) {
			return true
		}

		if(arguments == null || args == null) {
			return false
		}

		if(arguments.size != args.size) {
			return false
		}

		val count = arguments.size
		for(i in 0 until count) {
			if(arguments[i] != args[i]) {
				return false
			}
		}

		return true
	}

	fun doSuccess(data: Any?) {
		resultList.forEach {
			try {
				it.success(data)
			}
			catch (e: Throwable) {
				e.printStackTrace()
			}
		}
	}

	fun doError(errorCode: String, errorMessage: String?, errorDetail: Any?) {
		resultList.forEach {
			try {
				it.error(errorCode, errorMessage, errorDetail)
			}
			catch (e: Throwable) {

			}
		}
	}

	fun doNotImplemented() {
		resultList.forEach {
			try {
				it.notImplemented()
			}
			catch (e: Throwable) {

			}
		}
	}
}

File: ios/Classes/AsyncMethodProcessor.swift
//
//  AsyncMethodProcessor.swift
//  Runner
//
//  Created by CimZzz on 2020/3/15.
//  Copyright © 2020 The Chromium Authors. All rights reserved.
//

open class AsyncMethodProcessor : BaseMethodProcessor {
    private var childProcessorList = [AsyncProcessor].init()
    
    fileprivate func getAsyncProcessor(arguments: [NSObject]?) -> AsyncProcessor? {
        var processor: AsyncProcessor? = nil
        for existProcessor in childProcessorList {
            if(existProcessor.matchArguments(arguments)) {
                processor = existProcessor
                break
            }
        }

        return processor
    }
    
    fileprivate func deleteAsyncProcessor(processor: AsyncProcessor) {
        if let idx = childProcessorList.firstIndex(of: processor) {
            childProcessorList.remove(at: idx)
        }
    }

    open func asyncHandle(_ result: @escaping FlutterResult, _ arguments: [NSObject]? = nil, _ callback: @escaping (AsyncResultCallback) -> Void) {
        var processor = getAsyncProcessor(arguments: arguments)
        var isNewProcessor = false
        
        if(processor == nil) {
            isNewProcessor = true
            let newProcessor = AsyncProcessor(arguments: arguments)
            processor = newProcessor
            childProcessorList.append(newProcessor)
        }

        processor?.resultList.append(result)
        
        if(isNewProcessor) {
            let asyncCallback = AsyncResultCallback(methodProcessor: self, processor: processor!)
            callback(asyncCallback)
        }
    }
}

open class AsyncResultCallback: NSObject {
    private let methodProcessor: AsyncMethodProcessor
    private let processor: AsyncProcessor
    
    private var isCompleted = false
    /**
     * 0 - 成功
     * 1 - 失败
     * 2 - 未实现
     */
    private var completedMode = 0
    private var params: Any? = nil
    private var strParams1: String = ""
    private var strParams2: String? = nil
    
    fileprivate init(methodProcessor: AsyncMethodProcessor, processor: AsyncProcessor) {
        self.methodProcessor = methodProcessor
        self.processor = processor
        super.init()
    }
    
    open func doSuccess(data: Any?) {
        if(isCompleted) {
            return
        }
        isCompleted = true
        completedMode = 0
        params = data
        DispatchQueue.main.async {
            self.perform()
        }
    }

    open func doError(errorCode: String, errorMessage: String?, errorDetail: Any?) {
        if(isCompleted) {
            return
        }
        isCompleted = true
        completedMode = 1
        params = errorDetail
        strParams1 = errorCode
        strParams2 = errorMessage
        DispatchQueue.main.async {
            self.perform()
        }
    }


    open func doNotImplemented() {
        if(isCompleted) {
            return
        }
        isCompleted = true
        completedMode = 2
        DispatchQueue.main.async {
            self.perform()
        }
    }

    open func perform() {
        switch(completedMode) {
        case 0:
            processor.doSuccess(data: params)
            break
        case 1:
            processor.doError(errorCode: strParams1, errorMessage: strParams2, errorDetail: params)
            break
        case 2:
            processor.doNotImplemented()
            break
        default:
            break
        }
        
        methodProcessor.deleteAsyncProcessor(processor: processor)
    }
}

private class AsyncProcessor : NSObject {
    let arguments: [NSObject]?
    var resultList = [FlutterResult].init()
    
    init(arguments: [NSObject]?) {
        self.arguments = arguments
        super.init()
    }
    
    func matchArguments(_ args: [NSObject]?) -> Bool {
        if(arguments == nil && args == nil) {
            return true
        }

        if(arguments == nil || args == nil) {
            return false
        }
        
        if let selfArgs = self.arguments, let paramsArgs = args {
            if(selfArgs.count != paramsArgs.count) {
                return false
            }

            let count = selfArgs.count
            var i = 0;
            while(i < count) {
                if(selfArgs[i] != paramsArgs[i]) {
                    return false
                }
                i += 1
            }
            
            return true
        }
        
        return false
    }
    
    func doSuccess(data: Any?) {
        resultList.forEach {
            it in
            it(data)
        }
    }

    func doError(errorCode: String, errorMessage: String?, errorDetail: Any?) {
        resultList.forEach {
            it in
            it(FlutterError(code: errorCode, message: errorMessage, details: errorDetail))
        }
    }

    func doNotImplemented() {
        resultList.forEach {
            it in
            it(FlutterMethodNotImplemented)
        }
    }
}
